pkg/dynamic-config/dynamic_context.go

```diff
@@ -172,11 +172,12 @@ func (d *DynamicContext) HasLogin() bool {
 }
 
 func (d *DynamicContext) EnvironmentId() (string, error) {
-	if id := d.GetCurrentEnvironment(); id != "" {
-		return id, nil
+	id := d.GetCurrentEnvironment()
+	if _, err := d.V2Client.GetOrgEnvironment(id); err != nil {
```

Comment:
This gets used in nearly every Cloud command? Are we sure we want to introduce an extra API call to every command?

Reply:
I think it's worth it to prevent stale environment data from giving misleading errors. This particular bug sidesteps our suggestion to set an active environment, which adds to the confusion.

Alternatively, we could do this check if we run into a 403 from routes that take environments (and the user didn't use the environment flag).

Reply:
> Alternatively, we could do this check if we run into a 403 from routes that take environments

I'm conflicted on if this is the more correct approach here or not. This definitely falls in line with "optimize for 99% of cases"... I'm wondering if we could do this across the board, not just for environments but also for all of our other saved resources like Kafka clusters, etc.

Reply:
Yeah I don't see any reason why we couldn't. In hindsight there's nothing stopping users from running into this problem with other saved resources, and I think we've only seen this with environments so far because of how people are using CLI, UI, terraform (Brett confirmed that terraform destroy can cause this too).

Just to clarify: if I address all saved resources, do you think it would be the better approach then?

Reply:
> if I address all saved resources, do you think it would be the better approach then?

I think so. I think the only resources that we're saving properly right now are environment and Flink compute pool.

Reply:
(see `EnvironmentContext`)